core/src/main/java/hudson/tasks/Fingerprinter.java

```diff
@@ -390,7 +390,9 @@ public void add(Map<String,String> moreRecords) {
             Map<String,String> r = new HashMap<>(record);
             r.putAll(moreRecords);
             record = compact(r);
-            ref = null;
+            synchronized (this) {
+                ref = null;
+            }
```

Comment:
`ref` is manipulated elsewhere and is not volatile, so we should synchronize on `this` whenever reading or writing it.